### PR TITLE
Save screenshot as hash to resolve screenshot issue with the same name.

### PR DESCRIPTION
**Pull request by Guilty-Spark-343**
### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first.

### Description of the Change

Addresses the issue of graphs with the same name having the same screenshot on the welcome page.  In this implementation the graph screenshot is saved as the MD5 hash of the path instead of the filename.  If a graph with the same name is saved in two different directories it will display the correct screenshot instead of the most recent graph with the same name.  It is also backward compatible with the graph name, if it can't find the hash of the recent graph it will look for the graph name, failing that it will use the default.  

image

### Alternate Designs

We could have included the path in the screenshot name  but this could have had some weird side effects when saving the file.  We could have also included just the parent directory in the name but we may have had the same issue with the same name and same parent folder even if it was less likely. 

### Why Should This Be In Core?

Bugfix addressing #1499.

### Benefits

It better aligns what the user is expecting when they open the welcome page and gives a better representation of the recent graphs.

### Possible Drawbacks

If the graph is moved in the file system without saving in constellation then the path will no longer match the hash in the screenshots directory and will display the default icon instead of the screenshot. 

There is a very very small chance that the md5 hash collides with a path that isn't a match.  

### Verification Process

- Create a new graph
- Add a few nodes and save the graph
- Create a new graph
- Create a sphere graph (or something else, so long as the graph is different to the first one)
- Save the graph in a different location but with the same name as the previous graph
- Close and open Constellation
- The screenshots for each graph on the welcome page with the same name should match the graph it is opening instead of the same screenshot for all the graphs that have the same name.

### Applicable Issues

#1499

### CODE REVIEW

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java [hunk 1]
```diff
@@ -81,14 +85,53 @@ public static File getScreenshotsDir() {
...
+     * @return MD5 hash of filepath
+     */
+    private static String hashFilePath(final String filepath){
+        final MessageDigest md;
+        try {
+            md = MessageDigest.getInstance("MD5");
+            md.update(filepath.getBytes());
+            final byte[] digest = md.digest();
+            return DatatypeConverter.printHexBinary(digest).toUpperCase();            
+        } catch (NoSuchAlgorithmException ex) {
```
[the remark this answers is not in the document]
Comment (arcturus2) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java]: updated

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java [hunk 2]
```diff
@@ -81,14 +85,53 @@ public static File getScreenshotsDir() {
...
+            Exceptions.printStackTrace(ex);
```
[the remark this answers is not in the document]
Comment (arcturus2) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java]: updated

File: CoreWelcomeView/src/au/gov/asd/tac/constellation/views/welcome/WelcomeViewPane.java [hunk 3]
```diff
@@ -203,7 +203,7 @@ public WelcomeViewPane() {
                     recentGraphButtons[i].setTooltip(toolTip);
                     final String text = recentGraphButtons[i].getText();
 
-                    final String screenshotFilename = String.format(screenshotFilenameFormat, text);
+                    final String screenshotFilename = RecentGraphScreenshotUtilities.findScreenshot(fileDetails.get(i).getPath(), fileDetails.get(i).getFileName());
```
[the remark this answers is not in the document]
Comment (arcturus2) [reply CoreWelcomeView/src/au/gov/asd/tac/constellation/views/welcome/WelcomeViewPane.java]: I've changed `findScreenshot` to return an `Optional` as suggested and done some minor cleanup.

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java [hunk 4]
```diff
@@ -81,14 +85,53 @@ public static File getScreenshotsDir() {
...
+        return filepath;
+    }
+    
+    /**
+     * Finds the path for the screenshot
+     * @param filepath the filepath for the graph
+     * @param filename the filename of the graph
+     * @return the filepath to the screenshot png
+     */
+    public static String findScreenshot(final String filepath, final String filename){
```
[the remark this answers is not in the document]
Comment (arcturus2) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java]: Unit tests have been added.

File: toc.md [hunk 5]
```diff
@@ -2,176 +2,176 @@
 <div id="accordion">
 <div class="card"><div class="card-header"><h2 class="mb-0">Constellation Documentation</h2></div>
 <div class="card"><div class="card-header"><h2 class="mb-0"><button href="#" role="button" class="btn btn-link btn-block text-left collapsed" data-toggle="collapse" data-target="#Overview" aria-expanded="false" aria-controls="Overview">Overview</button></h2></div>
-<div id="Overview" class="collapse" aria-labelledby="Overview" data-parent="#Overview"> <div class="card-body">    <a href="..\constellation\CoreFunctionality\src\au\gov\asd\tac\constellation\functionality\docs\about-constellation.md">About Constellation</a><br/>
```
Comment (antares1470) [inline toc.md:5]: I think this was accidentally committed in. Are you able to revert the commit of this file?

Comment (arcturus2) [reply toc.md]: @antares1470 thanks. I have updated with the latest `master` and fixed the `toc.md` file.

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java [hunk 6]
```diff
@@ -82,14 +87,56 @@ public static File getScreenshotsDir() {
...
      * so that it can be used by the Welcome View.
      *
-     * @param filename The filename of the graph
+     * @param filepath The filepath of the graph
      */
-    public static void takeScreenshot(final String filename) {
-        final String imageFile = getScreenshotsDir() + File.separator + filename + FileExtensionConstants.PNG;
+    public static void takeScreenshot(final String filepath) {
+        final String pathHash = hashFilePath(filepath);
+        final String imageFile = getScreenshotsDir() + File.separator + pathHash + ".png";
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java:138]: `FileExtensionConstants.PNG`

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java [hunk 7]
```diff
@@ -166,10 +213,26 @@ public static void refreshScreenshotsDir() {
             filesInDirectory.addAll(Arrays.asList(screenShotsDir.listFiles()));
         }
 
-        RecentFiles.getUniqueRecentFiles().forEach(item -> filesInHistory.add(item.getFileName() + FileExtensionConstants.PNG));
+        RecentFiles.getUniqueRecentFiles().forEach(item -> {
+            filesInHistory.add(item.getFileName() + ".png");
```
Comment (antares1470) [inline CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java:217]: ditto

File: toc.md [hunk 8]
```diff
@@ -176,4 +176,4 @@
 </a> </div> </div> </div>
 </div>
 </div>
-</div>
\ No newline at end of file
+</div>
```
Comment (antares1470) [inline toc.md]: 😕

**Review by antares1470 — CHANGES_REQUESTED** [review]

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java (hunk 6 above)
Comment (arcturus2) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java]: Thanks @antares1470 - updated

File: CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java (hunk 7 above)
Comment (arcturus2) [reply CoreInteractiveGraph/src/au/gov/asd/tac/constellation/graph/interaction/plugins/io/screenshot/RecentGraphScreenshotUtilities.java]: Sorry I missed this one @antares1470. I also added another one in my most recent commit.

**Review by antares1470 — APPROVED** [review]

**Pull request merged**